src/shared/ui/pickerPrompter.ts

```diff
@@ -0,0 +1,255 @@
+    return new Promise<DataQuickPickItem<T>[] | undefined>(resolve => {
+        picker.onDidAccept(() => picker.selectedItems.length > 0 && resolve(Array.from(picker.selectedItems)))
+        picker.onDidHide(() => resolve(undefined))
+        picker.onDidTriggerButton(button => {
+            if (button === vscode.QuickInputButtons.Back) {
+                resolve([{ label: '', data: WIZARD_BACK }])
+            } else if ((button as QuickInputButton<T>).onClick !== undefined) {
+                const response = (button as QuickInputButton<T>).onClick!()
+                if (response !== undefined) {
+                    resolve([{ label: '', data: response }])
+                }
+            }
+        })
+        picker.show()
+    }).finally(() => picker.hide())
+}
+
+export class QuickPickPrompter<T> extends Prompter<T> {
+    protected _lastPicked?: DataQuickPickItem<T>
+
+    public set lastResponse(response: DataQuickPickItem<T> | undefined) {
+        if (response === undefined || !isDataQuickPickItem(response)) {
+            return
+        }
+
+        this.quickPick.activeItems = this.quickPick.items.filter(item => item.label === response.label)
+
+        if (this.quickPick.activeItems.length === 0) {
+            this.quickPick.activeItems = [this.quickPick.items[0]]
+        }
+    }
+
+    public get lastResponse() {
+        return this._lastPicked
+    }
+
+    constructor(public readonly quickPick: DataQuickPick<T>) {
+        super()
+    }
+
+    public setSteps(current: number, total: number): void {
+        this.quickPick.step = current
+        this.quickPick.totalSteps = total
+    }
+
+    protected async promptUser(): Promise<PromptResult<T>> {
+        const choices = await promptUser(this.quickPick)
+
+        if (choices === undefined) {
+            return choices
+        }
+
+        this._lastPicked = choices[0]
+        const result = choices[0].data
+
+        return result instanceof Function ? await result() : result
+    }
+
+    public setLastResponse(picked: DataQuickPickItem<T> | undefined = this._lastPicked): void {
+        if (picked === undefined) {
+            return
+        }
+
+        this.quickPick.activeItems = this.quickPick.items.filter(item => item.label === picked.label)
+
+        if (this.quickPick.activeItems.length === 0) {
+            this.quickPick.activeItems = [this.quickPick.items[0]]
+        }
+    }
+}
+
+/**
+ * Allows the prompter to accept the QuickPick filter box as input, shown as a QuickPickItem.
+ * Note: untested with any picker that can refresh items (haven't needed this functionality yet)
+ *
+ * @param label The label of the QuickPickItem that shows the user's input
+ * @param transform Required when the expected type is not a string, transforming the input into the expected type or a control signal.
+ */
+export class CustomQuickPickPrompter<T> extends QuickPickPrompter<T> {
```

Comment:
rather than a separate type, could this be just an optional ctor arg on QuickPickPrompter? That may be easier to understand, and discover. (And avoids awkward "CustomFoo" name, which raises the question "aren't all of my wizards/prompters/pickers custom?")

Reply:
Yes options _could_ be used, but it starts to add a lot of additionally complexity in the base class that I think is best kept separate. I agree that though that the naming is bad. I just can't think of a good name for the functionality of having a QuickPick accept arbitrary user-input through the filter-box.

`UserInputQuickPick`? `FilterBoxInputQuickPick`? I guess a more verbose name is acceptable since the class is extremely specific.

Reply:
> `UserInputQuickPick`? `FilterBoxInputQuickPick`? I guess a more verbose name is acceptable since the class is extremely specific.

Those could be better. This logic is fairly isolated though, so an optional ctor param doesn't seem too bad to me.